runtime/caml/config.h

```diff
@@ -223,7 +223,7 @@ typedef uint64_t uintnat;
 /* Maximum size of the minor zone (words).
    Must be greater than or equal to [Minor_heap_min].
 */
-#define Minor_heap_max (1 << 28)
+#define Minor_heap_max (1 << 21)
```

Comment:
This is very small (16MB); Coq uses 256MB by default. Instead, the reservation of contiguous address space for minor heaps is a remnant of the concurrent collector which should probably be removed.

Reply:
OK, thanks for the feedback.
After opening the PR I learned that the increase from ade8a06 is intentional.
I expect #733 will track this going forward (and that this will be closed).

Reply:
> the reservation of contiguous address space for minor heaps is a remnant of the concurrent collector which should probably be removed.

@gadmm for some additional context, we experimented with domain-local allocation buffers (DLABs), where a user-configurable minor heap area is allocated by a single domain and each domain carves a page for its own use. This implementation turned out to suffer from severe memory hierarchy impact (suspect both NUMA and cache effects). So performance-wise, it turned out to be sensible to retain the current implementation. See the report here: https://github.com/ocaml-multicore/ocaml-multicore/wiki/Domain-Local-Allocation-Buffers-Addendum. 

In terms of desiderata, any alternative design we come up will need to have:

1. A fast `Is_minor` check so that we can very cheaply check whether a value is in the minor heap. This is true in the current design since all of the minor heap arenas are in a reserved contiguous segment. `Is_minor` check needs only two comparisons. 
2. A minor heap allocation strategy where each domain allocates its own memory. This is true in the current design since the primary domain only reserves the contiguous space, but the individual domains themselves commit the memory. This turns out to be quite important as the DLAB work shows.

Reply:
@kayceesrk Thanks for the additional context. I am not sure to understand what was the motivation behind DLABs (the expected benefits) and I cannot tell what is going on from the description. The design I suggested at #733 is equivalent to the current one in terms of the domain allocating and keeping its own memory, so I expect that it avoids these issues. (But if the tail cost of spawning a domain is not an issue, then growing the contiguous space geometrically after a STW phase sounds easier in short term, of course.)

Reply:
@kayceesrk I now read more about DLABs and better understand how they reduced virtual memory consumption. If I understand correctly, #508 did not yet implement non-fixed minor heap sizes, so we know for sure that there is no impact of varying minor heap sizes on the rare observed speedup / some of the slowdown?

In addition to lower virtual memory usage, there is a benefit of allocating (not merely reserving) as contiguously as possible (i.e. avoiding big holes of unmapped memory) in terms of cache interference (according to the Streamflow paper) and MMU cache locality (in my experiment), which both DLABs and my design should have over the current one. I understand the wall you hit with DLABs. To sum up, implementing `Is_young` with a bibop is a way to open the design space by not requiring the minor heaps to occupy a fixed contiguous area. But I agree that it makes sense to try again with the DLABs design as suggested by @ctk21 in a way that respects your point 2. (smaller investment, higher cost of spawning a domain, solves a smaller problem).

Thanks for mentioning NUMA. The page allocator in my experiment should certainly be made NUMA-aware (but providing such control is what a page allocator is there for after all).

Reply:
The point of DLABs, which is not mentioned in the description, is to address a form of early promotion in the current design. In the current design, we trigger a minor collection when any one of the domains exhausts its minor heap arena. If we have `N` domains, each with `S` words of the minor heap space, it is likely that we will trigger a minor GC when the used space in the minor heap is much less than `N * S`. (We haven't empirically measured this, and we should if we are reviving the DLABs work). In the worst case, we will trigger a minor GC when one of the domains fills up its minor heap arena and others are empty -- worst-case utilised space is `S`. 

The idea of DLABs was to get close to `N * S` utilization by making the domains request pages of size `P` from the shared minor heap space. In the worst case, a minor GC will be triggered when one domain has filled up the last available page and all the others haven't started filling their last requested pages. So `N * S - (N-1) * P`. Given that `P << S`, this would be a win in terms of preventing early promotion. This was the reasoning behind going for DLABS. 

> Thanks for mentioning NUMA. The page allocator in my experiment should certainly be made NUMA-aware (but providing such control is what a page allocator is there for after all).

You will get NUMA friendliness on Linux by default if you allocate memory from the thread that will use it.

We should discuss the possibility of adapting bibop for multicore minor heap in https://github.com/ocaml-multicore/ocaml-multicore/issues/733 in order to avoid splintering the discussions across several threads. I had some questions here: https://github.com/ocaml-multicore/ocaml-multicore/issues/733#issuecomment-963807435.

Reply:
(About DLABs; I will discuss bibop elsewhere.)

I understood later. With the current multicore design it is hard to avoid false promotions, and it sounds better to collect when a fixed amount has been allocated (but this does not entirely avoid premature collections). This idea could be tried with a page allocator (minor heap extents could be pooled per core or per-NUMA-node).

> You will get NUMA friendliness on Linux by default if you allocate memory from the thread that will use it.

I expect it affects how I implement reclamation (`MADV_FREE` would probably not be enough).

Reading the DLABs discussions (esp. the commit_decommit variant) I also wanted to mention that you can map the whole mapping once in a thread, instead of mapping N times where N is the number of domains, since what counts for node attribution is who first touches the page.